File: android/src/main/kotlin/com/fingerprintjs/flutter/fpjs_pro/fpjs_pro_plugin/FingerprintHostApiImpl.kt
// Pigeon HostApi: identification via Android Fingerprint SDK 4.x.
package com.fingerprintjs.flutter.fpjs_pro.fpjs_pro_plugin

import android.content.Context
import com.fingerprint.android.Configuration
import com.fingerprint.android.FingerprintResponse

internal class FingerprintHostApiImpl(
  private val applicationContext: Context,
  private val clientCache: FingerprintClientCache = FingerprintClientCache(applicationContext),
) : FingerprintHostApi {

  override fun create(config: FingerprintNativeConfig) {
    nativeClient(config)
  }

  override fun get(
    config: FingerprintNativeConfig,
    tags: Map<String?, Any?>?,
    linkedId: String?,
    timeoutMs: Long?,
    callback: (Result<FingerprintNativeResult>) -> Unit,
  ) {
    val client = try {
      nativeClient(config)
    } catch (error: FlutterError) {
      callback(Result.failure(error))
      return
    }
    val tagMap = pigeonTagsToNative(tags)
    val linked = linkedId ?: ""
    val listener: (FingerprintResponse) -> Unit = { response ->
      callback(
        Result.success(
          // Native FingerprintResponse -> Pigeon result.
          FingerprintNativeResult(
            eventId = response.eventId,
            visitorId = response.visitorId,
            suspectScore = response.suspectScore?.toLong(),
            sealedResult = response.sealedResult,
          ),
        ),
      )
    }
    val errorListener: (com.fingerprint.android.Error) -> Unit = { error ->
      callback(Result.failure(toFlutterError(error)))
    }
    if (timeoutMs != null) {
      client.getVisitorId(timeoutMs.toInt(), tagMap, linked, listener, errorListener)
    } else {
      client.getVisitorId(tagMap, linked, listener, errorListener)
    }
  }

  private fun nativeClient(config: FingerprintNativeConfig) =
    clientCache.getOrCreate(buildConfiguration(config), config.pluginVersion)

  private fun buildConfiguration(config: FingerprintNativeConfig): Configuration {
    val region = parseRegion(config.region)
    // Empty primary still uses the region URL. Fallbacks are independent, so
    // they survive when only endpointFallbacks is set.
    // https://docs.fingerprint.com/docs/android-sdk
    val endpointUrl = config.endpoint?.takeIf { it.isNotEmpty() } ?: region.endpointUrl
    val fallbacks = config.endpointFallbacks?.filter { it.isNotEmpty() } ?: emptyList()
    val locationTimeout = config.locationTimeoutMillis ?: 5000L
    return Configuration(
      config.apiKey,
      region,
      endpointUrl,
      fallbacks,
      listOf(Pair("fingerprint-pro-flutter", config.pluginVersion)),
      config.allowUseOfLocationData,
      locationTimeout,
    )
  }
}

internal fun parseRegion(region: String?): Configuration.Region {
  return when (region?.lowercase()) {
    "eu" -> Configuration.Region.EU
    "ap" -> Configuration.Region.AP
    "us", null -> Configuration.Region.US
    else -> throw FlutterError("unknown_error", "Invalid region: $region", null)
  }
}

// Keep JSON null values. The SDK type is Map<String, Any>, so this is an
// unchecked cast of a HashMap that may contain nulls. Null keys cannot
// be stored.
// https://docs.fingerprint.com/docs/tagging-information
internal fun pigeonTagsToNative(tags: Map<String?, Any?>?): Map<String, Any> {
  if (tags == null) {
    return emptyMap()
  }
  val result = HashMap<String, Any?>(tags.size)
  for ((key, value) in tags) {
    if (key != null) {
      result[key] = value
    }
  }
  @Suppress("UNCHECKED_CAST")
  return result as Map<String, Any>
}

File: ios/fpjs_pro_plugin/Sources/fpjs_pro_plugin/FingerprintHostApiImpl.swift
// Pigeon HostApi: identification via iOS Fingerprint SDK 4.x.

@preconcurrency import Fingerprint
import Foundation

// Pigeon completion is not Sendable. Box it so the SDK callback can call it
// under Swift 6.
// https://docs.swift.org/swift-book/documentation/the-swift-programming-language/concurrency/#Sendable-Types
private final class CompletionBox: @unchecked Sendable {
  let completion: (Result<FingerprintNativeResult, Error>) -> Void

  init(_ completion: @escaping (Result<FingerprintNativeResult, Error>) -> Void) {
    self.completion = completion
  }
}

/// Pigeon HostApi backed by the iOS Fingerprint SDK 4.x.
final class FingerprintHostApiImpl: FingerprintHostApi {
  private let clientCache: FingerprintClientCache

  init(clientCache: FingerprintClientCache = FingerprintClientCache()) {
    self.clientCache = clientCache
  }

  func create(config: FingerprintNativeConfig) throws {
    _ = nativeClient(for: config)
  }

  func get(
    config: FingerprintNativeConfig,
    tags: [String?: Any?]?,
    linkedId: String?,
    timeoutMs: Int64?,
    completion: @escaping (Result<FingerprintNativeResult, Error>) -> Void
  ) {
    let client = nativeClient(for: config)
    let metadata = prepareMetadata(linkedId: linkedId, tags: tags)
    let timeoutSeconds: TimeInterval? = timeoutMs.map { TimeInterval($0) / 1000.0 }
    let box = CompletionBox(completion)

    let handler: VisitorIdResponseBlock = { result in
      switch result {
      case .success(let response):
        box.completion(
          .success(
            // Native FingerprintResponse -> Pigeon result.
            FingerprintNativeResult(
              eventId: response.eventId,
              visitorId: response.visitorId,
              suspectScore: response.suspectScore.map { Int64($0) },
              sealedResult: response.sealedResult
            )
          )
        )
      case .failure(let error):
        let fields = error.pigeonFields()
        box.completion(
          .failure(PigeonError(code: fields.code, message: fields.message, details: fields.eventId))
        )
      }
    }

    if let timeoutSeconds {
      client.getVisitorIdResponse(metadata, timeout: timeoutSeconds, completion: handler)
    } else {
      client.getVisitorIdResponse(metadata, completion: handler)
    }
  }

  private func nativeClient(for config: FingerprintNativeConfig) -> FingerprintClientProviding {
    clientCache.getOrCreate(
      configuration: buildConfiguration(config: config),
      pluginVersion: config.pluginVersion
    )
  }

  private func buildConfiguration(config: FingerprintNativeConfig) -> Configuration {
    let fallbacks = config.endpointFallbacks?.filter { !$0.isEmpty } ?? []
    let namedRegion = Self.namedRegion(config.region)
    let region: Region
    if let endpoint = config.endpoint, !endpoint.isEmpty {
      region = .custom(domain: endpoint, fallback: fallbacks)
    } else if !fallbacks.isEmpty {
      // iOS only attaches fallbacks on Region.custom. With no primary URL, use
      // the region's default endpoint as that custom domain so fallbacks still
      // apply. Android does the same with `endpoint ?: region.endpointUrl`.
      // https://docs.fingerprint.com/docs/ios-sdk
      region = .custom(domain: Self.defaultEndpoint(namedRegion), fallback: fallbacks)
    } else {
      region = namedRegion
    }
    return Configuration(
      apiKey: config.apiKey,
      region: region,
      integrationInfo: [("fingerprint-pro-flutter", config.pluginVersion)],
      allowUseOfLocationData: config.allowUseOfLocationData
    )
  }

  private func prepareMetadata(linkedId: String?, tags: [String?: Any?]?) -> Metadata {
    var metadata = Metadata(linkedId: linkedId)
    guard let tags else {
      return metadata
    }
    for (key, value) in tags {
      guard let key else { continue }
      if let converted = jsonType(from: value) {
        metadata.setTag(converted, forKey: key)
      }
    }
    return metadata
  }

  private static func namedRegion(_ region: String?) -> Region {
    switch region?.lowercased() {
    case "eu":
      return .eu
    case "ap":
      return .ap
    default:
      return .global
    }
  }

  private static func defaultEndpoint(_ region: Region) -> String {
    switch region {
    case .eu:
      return "https://eu.api.fpjs.io"
    case .ap:
      return "https://ap.api.fpjs.io"
    default:
      return "https://api.fpjs.io"
    }
  }
}
